File: backend/app/models/study_day.py
"""StudyDay ORM model."""

from __future__ import annotations

from datetime import date, datetime
from typing import TYPE_CHECKING, Optional

from sqlalchemy import (
    Date,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy import Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql.functions import now

from app.database import Base
from app.models.enums import StudyStatus

if TYPE_CHECKING:
    from app.models.daily_tracker import DailyTracker
    from app.models.plan import Plan


class StudyDay(Base):
    """A single day within a plan (Day 1 .. Day N)."""

    __tablename__ = "study_days"
    __table_args__ = (
        UniqueConstraint("plan_id", "day_number", name="uq_study_day_plan_day"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    plan_id: Mapped[int] = mapped_column(
        ForeignKey("plans.id", ondelete="CASCADE"), nullable=False, index=True
    )

    day_number: Mapped[int] = mapped_column(Integer, nullable=False)
    week_number: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    phase: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    focus: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    task: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    difficulty: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)

    status: Mapped[StudyStatus] = mapped_column(
        SAEnum(StudyStatus, name="study_status"),
        nullable=False,
        default=StudyStatus.NOT_STARTED,
    )

    scheduled_date: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    completed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=now(),
        onupdate=now(),
        nullable=False,
    )

    plan: Mapped["Plan"] = relationship(back_populates="days")
    tracker: Mapped[Optional["DailyTracker"]] = relationship(
        back_populates="study_day",
        cascade="all, delete-orphan",
        uselist=False,
    )

File: frontend/lib/models/study_day.dart
import 'study_status.dart';

class StudyDay {
  final int id;
  final int planId;
  final int dayNumber;
  final int? weekNumber;
  final String? phase;
  final String? focus;
  final String? task;
  final String? difficulty;
  final StudyStatus status;
  final DateTime? scheduledDate;
  final DateTime? completedAt;
  final String? notes;

  const StudyDay({
    required this.id,
    required this.planId,
    required this.dayNumber,
    required this.status,
    this.weekNumber,
    this.phase,
    this.focus,
    this.task,
    this.difficulty,
    this.scheduledDate,
    this.completedAt,
    this.notes,
  });

  factory StudyDay.fromJson(Map<String, dynamic> json) {
    return StudyDay(
      id: json['id'] as int,
      planId: json['plan_id'] as int,
      dayNumber: json['day_number'] as int,
      weekNumber: json['week_number'] as int?,
      phase: json['phase'] as String?,
      focus: json['focus'] as String?,
      task: json['task'] as String?,
      difficulty: json['difficulty'] as String?,
      status: StudyStatus.fromWire(json['status'] as String?),
      scheduledDate: _date(json['scheduled_date']),
      completedAt: _date(json['completed_at']),
      notes: json['notes'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'plan_id': planId,
        'day_number': dayNumber,
        'week_number': weekNumber,
        'phase': phase,
        'focus': focus,
        'task': task,
        'difficulty': difficulty,
        'status': status.wire,
        'scheduled_date': scheduledDate?.toIso8601String(),
        'completed_at': completedAt?.toIso8601String(),
        'notes': notes,
      };

  static DateTime? _date(dynamic value) {
    if (value == null) return null;
    return DateTime.tryParse(value.toString());
  }
}
